<b>`LONDON SMART METER DATA ANALYSIS`</b>

There are `5` important files in this dataset :

<b>`informations_households.csv`</b> : This file that contains all the information on the households in the panel (their acorn group, their tariff) and in which block.csv.gz file their data are stored.

<b>`halfhourly_dataset` </b>: These files that contain the block files with the half-hourly smart meter measurement

<b>`daily_dataset`</b>: These files that contain the block files with the daily information like the number of measures, minimum, maximum, mean, median, sum and std.

<b>`acorn_details.csv`</b> : Details on the acorn groups and their profile of the people in the group.

<b>`weatherdailydarksky.csv`</b> : This contains the daily data from darksky api. 

Our analysis makes use of all these files to bring out insights

In [0]:
# Importing all required libraries
import pandas as pd
from pyspark import SparkContext
from pyspark.sql import SQLContext
from pyspark.sql.session import SparkSession
from pyspark.sql.functions import *
import seaborn as sns
import pandas as pd
import matplotlib.pyplot as plt

In [0]:
# Importing all blocks from Daily_Dataset Folder into single dataframe

daily_df_final = pd.DataFrame()

for i in range(0,111):
    path = 'https://raw.githubusercontent.com/OmkarKatkade/Data-603/main/daily_dataset/daily_dataset/block_' + str(i) + '.csv';
    daily_df = pd.read_csv(path)
    daily_df_final = daily_df_final.append(daily_df)
    
daily_df_final
df = spark.createDataFrame(daily_df_final)

In [0]:
#Importing Household data

household = pd.read_csv("https://raw.githubusercontent.com/OmkarKatkade/Data-603/main/informations_households.csv", encoding= 'unicode_escape')
household = spark.createDataFrame(household)
household.show()

+---------+--------+-------+-------------+-------+
|    LCLid|stdorToU|  Acorn|Acorn_grouped|   file|
+---------+--------+-------+-------------+-------+
|MAC005492|     ToU| ACORN-|       ACORN-|block_0|
|MAC001074|     ToU| ACORN-|       ACORN-|block_0|
|MAC000002|     Std|ACORN-A|     Affluent|block_0|
|MAC003613|     Std|ACORN-A|     Affluent|block_0|
|MAC003597|     Std|ACORN-A|     Affluent|block_0|
|MAC003579|     Std|ACORN-A|     Affluent|block_0|
|MAC003566|     Std|ACORN-A|     Affluent|block_0|
|MAC003557|     Std|ACORN-A|     Affluent|block_0|
|MAC003553|     Std|ACORN-A|     Affluent|block_0|
|MAC003482|     Std|ACORN-A|     Affluent|block_0|
|MAC003463|     Std|ACORN-A|     Affluent|block_0|
|MAC003449|     Std|ACORN-A|     Affluent|block_0|
|MAC003428|     Std|ACORN-A|     Affluent|block_0|
|MAC003423|     Std|ACORN-A|     Affluent|block_0|
|MAC003422|     Std|ACORN-A|     Affluent|block_0|
|MAC003400|     Std|ACORN-A|     Affluent|block_0|
|MAC003394|     Std|ACORN-A|   

In [0]:
# Merging household data with daily dataset using inner join on 'LCLid'
df = df.join(household, ['LCLid'], how='inner')
df.show()

+---------+----------+------------------+------------------+----------+------------+------------------+------------------+------------------+--------+-------+-------------+-------+
|    LCLid|       day|     energy_median|       energy_mean|energy_max|energy_count|        energy_std|        energy_sum|        energy_min|stdorToU|  Acorn|Acorn_grouped|   file|
+---------+----------+------------------+------------------+----------+------------+------------------+------------------+------------------+--------+-------+-------------+-------+
|MAC004029|2012-07-06|            0.4625|0.5432916666666666|      1.17|          24|0.3022701756011846|            13.039|             0.096|     Std|ACORN-C|     Affluent|block_5|
|MAC004029|2012-07-07|            0.2255|          0.253125|     0.696|          48|0.1583076220258815|             12.15|             0.049|     Std|ACORN-C|     Affluent|block_5|
|MAC004029|2012-07-08|0.1654999999999999|0.2285416666666666|      0.84|          48|0.182216840

<b>Data Cleaning</b>

In [0]:
#Dropping the 'energy_std' column as it is less relevant to our analysis
df = df.drop('energy_std')

In [0]:
#Dropping rows with nulls values
df = df.dropna()

In [0]:
#Lets see the count of rows based on Acorn Group
count1 = df.groupBy('Acorn_grouped').count().orderBy('count')
display(count1)

Acorn_grouped count ACORN- 781 ACORN-U 19644 Comfortable 937061 Adversity 1135331 Affluent 1407678

INFERENCE: Group "ACORN-" and "ACORN-U" are significantly lower in terms of no. of records. Therefore we can ignore these two groups for our analysis.

In [0]:
# Dropping 'ACORN-' and 'ACORN-U' groups from the dataset.
df = df.filter(df.Acorn_grouped != 'ACORN-')
df = df.filter(df.Acorn_grouped != 'ACORN-U')

<b>Exploratory Data Analysis</b>

<b> `1. YEAR WISE ENERGY CONSUMPTION TREND BASED ON ACORN GROUP `</b>

In [0]:
'''
Extracting Day of the week, Month and Year from the 'day' column.
This will help in aggregation going forward
'''

df = df.withColumn('dayOfWeek', dayofweek(col('day')))
df = df.withColumn('month', month(col('day')))
df = df.withColumn('year', year(col('day')))
df.printSchema()

root
 |-- LCLid: string (nullable = true)
 |-- day: string (nullable = true)
 |-- energy_median: double (nullable = true)
 |-- energy_mean: double (nullable = true)
 |-- energy_max: double (nullable = true)
 |-- energy_count: long (nullable = true)
 |-- energy_sum: double (nullable = true)
 |-- energy_min: double (nullable = true)
 |-- stdorToU: string (nullable = true)
 |-- Acorn: string (nullable = true)
 |-- Acorn_grouped: string (nullable = true)
 |-- file: string (nullable = true)
 |-- dayOfWeek: integer (nullable = true)
 |-- month: integer (nullable = true)
 |-- year: integer (nullable = true)



In [0]:
'''
Now that we have Day of the week, Month and Year
extracted from the 'day' column, we can go ahead and check
year-wise energy consumption based on acorn group.
'''

t1 = df.groupBy("year", "Acorn_grouped").sum("energy_sum").withColumnRenamed("sum(energy_sum)", "Total_Energy_Consumption(kWh)").collect()
display(t1)

year Acorn_grouped Total_Energy_Consumption(kWh) 2012 Affluent 5905451.254876445 2014 Affluent 1496936.1979974064 2011 Affluent 46217.2389995 2013 Affluent 8709586.250781491 2014 Comfortable 904786.9739640976 2012 Comfortable 3109564.0009666 2011 Comfortable 28142.78900109998 2013 Comfortable 5332295.709879235 2011 Adversity 28592.354996600014 2012 Adversity 3353160.413988029 2013 Adversity 5373100.987924468 2014 Adversity 899299.8580024024

<b>`INFERENCE:` </b> Since we have significantly lower number of records for the year 2011 and 2014 and therefore energy consumption for these years is shown significantly lower than other years in the dataset. It is also important to notice that the households belonging to the Acorn Group 'Affluent' tend to consume the most electricity. Our initial assumption was that the group 'Comfortable' would be consumping a lot more electricity than the group 'Adversity' but it turns out that both these groups consume almost equal amount of electricity.

<b> `2. MONTHLY ENERGY CONSUMPTION BASED ON ACORN GROUP `</b>

In [0]:
'''
Now that we have seen the yearly trend of energy consumption,
we are curious to see the monthly trend of energy consumption across 
these acorn groups. 
'''
t2 = df.groupBy("month", "Acorn_grouped").sum("energy_sum").withColumnRenamed("sum(energy_sum)", "Total_Energy_Consumption(kWh)").collect()

In [0]:
display(t2)

month Acorn_grouped Total_Energy_Consumption(kWh) 2 Affluent 1705632.3189626068 1 Affluent 1900032.3069310065 4 Affluent 998308.6029894009 11 Affluent 1649664.247970792 10 Affluent 1436340.5169611035 8 Affluent 1084161.1579873997 6 Affluent 1059878.7469981017 3 Affluent 1136309.5819712994 5 Affluent 1047072.5709592995 9 Affluent 1197749.6989761982 12 Affluent 1836836.8979687071 7 Affluent 1106204.2939789004 9 Comfortable 691324.0919888039 7 Comfortable 663944.7499900996 6 Comfortable 607274.3019954001 3 Comfortable 614357.5059664014 1 Comfortable 1110102.1119789009 8 Comfortable 657486.220994798 10 Comfortable 857168.5239896041 11 Comfortable 965958.373966103 12 Comfortable 1109277.6509861003 5 Comfortable 580764.3429899997 4 Comfortable 538908.8529868993 2 Comfortable 978222.7459779958 7 Adversity 716995.8239908032 2 Adversity 971793.4009920035 11 Adversity 965753.8109919047 5 Adversity 635137.8189899991 4 Adversity 568438.3330149013 6 Adversity 660881.672994802 12 Adversity 1095475.6229715971 3 Adversity 612711.0300020005 1 Adversity 1105504.9949934955 8 Adversity 712317.2479940018 9 Adversity 732098.574973497 10 Adversity 877045.2830025037

<b>`INFERENCE:` </b> It is quite evident from the above visualization that households from the Affluent group tend to consume maximum electricity irrespective of the month. Additionally, it is also important to notice that the month of <b>January</b> is when the energy consumption is at the peak across all acorn groups. The Affluent group consumped around <b> 1.9 million kWh </b> of energy in the month of January across all years in the dataset.

<b> `3. DAILY ENERGY CONSUMPTION`</b>

In [0]:
'''
Let's see the trend of the total energy consumption on daily basis
'''

# energy_sum = df.select("day", "energy_sum").where(df.Acorn_grouped == 'Affluent')
energy_sum_0 = df.groupBy('day', 'LCLid').avg('energy_sum').withColumnRenamed("avg(energy_sum)", "Avg_Energy_Consumption(kWh)").collect()
display(energy_sum_0)

day LCLid Avg_Energy_Consumption(kWh) 2012-07-06 MAC004029 13.039 2012-07-07 MAC004029 12.15 2012-07-08 MAC004029 10.97 2012-07-09 MAC004029 13.679000000000002 2012-07-10 MAC004029 20.203 2012-07-11 MAC004029 12.423 2012-07-12 MAC004029 13.437 2012-07-13 MAC004029 20.385 2012-07-14 MAC004029 11.707 2012-07-15 MAC004029 17.380000000000003 2012-07-16 MAC004029 16.092000000000002 2012-07-17 MAC004029 19.125000000000004 2012-07-18 MAC004029 12.773 2012-07-19 MAC004029 11.484 2012-07-20 MAC004029 14.833 2012-07-21 MAC004029 10.884 2012-07-22 MAC004029 10.598000000000004 2012-07-23 MAC004029 13.591 2012-07-24 MAC004029 15.339999999999998 2012-07-25 MAC004029 8.317 2012-07-26 MAC004029 11.04 2012-07-27 MAC004029 15.094 2012-07-28 MAC004029 12.032000000000002 2012-07-29 MAC004029 14.024 2012-07-30 MAC004029 13.2880001 2012-07-31 MAC004029 17.1549998 2012-08-01 MAC004029 11.283000000000005 2012-08-02 MAC004029 11.796999999999995 2012-08-03 MAC004029 14.599000000000004 2012-08-04 MAC004029 14.094 2012-08-05 MAC004029 10.063000000000004 2012-08-06 MAC004029 10.217999999999998 2012-08-07 MAC004029 13.121000000000002 2012-08-08 MAC004029 13.002999999999998 2012-08-09 MAC004029 9.422 2012-08-10 MAC004029 7.187000000000002 2012-08-11 MAC004029 4.435000000000002 2012-08-12 MAC004029 4.349000000000001 2012-08-13 MAC004029 4.175000000000001 2012-08-14 MAC004029 4.279 2012-08-15 MAC004029 4.3450000000000015 2012-08-16 MAC004029 4.177000000000001 2012-08-17 MAC004029 6.663 2012-08-18 MAC004029 13.963999999999997 2012-08-19 MAC004029 15.057000000000002 2012-08-20 MAC004029 9.974000000000002 2012-08-21 MAC004029 14.953 2012-08-22 MAC004029 8.972 2012-08-23 MAC004029 10.638 2012-08-24 MAC004029 14.087999999999996 2012-08-25 MAC004029 14.337 2012-08-26 MAC004029 11.391 2012-08-27 MAC004029 11.367 2012-08-28 MAC004029 12.275999999999998 2012-08-29 MAC004029 10.904 2012-08-30 MAC004029 9.711 2012-08-31 MAC004029 15.3020001 2012-09-01 MAC004029 12.389 2012-09-02 MAC004029 9.304000000000002 2012-09-03 MAC004029 9.428999999999998 2012-09-04 MAC004029 16.121999999999996 2012-09-05 MAC004029 7.964000000000001 2012-09-06 MAC004029 8.972000000000001 2012-09-07 MAC004029 13.914999999999996 2012-09-08 MAC004029 13.714 2012-09-09 MAC004029 9.575999999999995 2012-09-10 MAC004029 8.805000000000001 2012-09-11 MAC004029 13.713 2012-09-12 MAC004029 7.948 2012-09-13 MAC004029 8.228000000000002 2012-09-14 MAC004029 15.463 2012-09-15 MAC004029 12.107999999999995 2012-09-16 MAC004029 8.666 2012-09-17 MAC004029 11.584000000000003 2012-09-18 MAC004029 13.901 2012-09-19 MAC004029 6.638999999999999 2012-09-20 MAC004029 7.785 2012-09-21 MAC004029 10.557 2012-09-22 MAC004029 8.955999999999998 2012-09-23 MAC004029 13.024 2012-09-24 MAC004029 9.325 2012-09-25 MAC004029 14.512 2012-09-26 MAC004029 10.862000000000002 2012-09-27 MAC004029 10.251 2012-09-28 MAC004029 12.297999999999998 2012-09-29 MAC004029 13.759 2012-09-30 MAC004029 10.427000000000003 2012-10-01 MAC004029 10.342999999999998 2012-10-02 MAC004029 15.195 2012-10-03 MAC004029 8.545 2012-10-04 MAC004029 9.363 2012-10-05 MAC004029 13.006 2012-10-06 MAC004029 10.657 2012-10-07 MAC004029 11.208000000000002 2012-10-08 MAC004029 8.651 2012-10-09 MAC004029 14.683999999999996 2012-10-10 MAC004029 9.77 2012-10-11 MAC004029 9.070000000000002 2012-10-12 MAC004029 12.280000000000005 2012-10-13 MAC004029 14.634000000000002 2012-10-14 MAC004029 18.393000000000004 2012-10-15 MAC004029 11.13 2012-10-16 MAC004029 15.634 2012-10-17 MAC004029 10.465 2012-10-18 MAC004029 8.849999999999998 2012-10-19 MAC004029 12.211 2012-10-20 MAC004029 12.0740001 2012-10-21 MAC004029 11.21 2012-10-22 MAC004029 13.257999999999996 2012-10-23 MAC004029 15.1 2012-10-24 MAC004029 6.988000000000001 2012-10-25 MAC004029 8.443 2012-10-26 MAC004029 13.61 2012-10-27 MAC004029 15.7939999 2012-10-28 MAC004029 15.887000000000002 2012-10-29 MAC004029 9.761 2012-10-30 MAC004029 15.98 2012-10-31 MAC004029 10.004 2012-11-01 MAC004029 10.608 2012-11-02 MAC004029 14.463000

<b>`INFERENCE:` </b> The total energy consumption reaches the peak in January 2013 with the value of 81.2 thousand kWh.

<b> `4. How much energy on an average does a typical household in UK consume in a day across the year. `</b>

In [0]:
'''
It is also important to know the average of how much energy does a typical household in UK consume in a day across the year.
'''
energy_sum_0 = spark.createDataFrame(energy_sum_0)
energy_sum_0 = energy_sum_0.groupBy('day').avg('Avg_Energy_Consumption(kWh)').collect()
display(energy_sum_0)

day avg(Avg_Energy_Consumption(kWh)) 2013-03-14 11.75087685989583 2014-02-16 11.580864293667071 2014-02-22 10.771510681257562 2012-10-21 10.942398086705422 2012-03-04 13.475282051165502 2013-08-20 7.657049635396703 2013-12-28 11.37412870699881 2012-03-09 12.162531287156973 2012-01-12 12.062221444755243 2013-11-08 10.527465251352927 2012-08-17 7.7477993645974585 2012-09-05 8.122021798835979 2013-09-27 8.522007738343978 2012-05-27 8.447817680463766 2012-05-30 7.971275208022282 2012-07-06 8.197206355891034 2012-11-08 10.926872671867006 2013-02-17 12.001773765198973 2014-02-08 11.562959951762114 2012-02-20 13.339136367245988 2012-01-28 13.37329690601093 2013-05-24 9.541277536189044 2012-03-19 11.04867480949248 2012-05-12 9.152971524840764 2011-12-16 13.416753835 2012-11-25 12.162683635449737 2013-02-12 12.898422186846945 2013-11-05 10.70833761213535 2012-06-15 7.224001546338319 2012-09-14 8.277171812455364 2012-10-06 9.545797282249803 2013-02-09 12.823108187930083 2013-10-05 9.151966097074776 2013-10-29 9.915919385606056 2012-08-24 7.8781958498835465 2013-04-09 10.883220001607473 2012-04-16 10.458688557737318 2012-09-03 8.333367294450325 2013-09-29 9.44141956722738 2013-10-07 8.945738126245391 2011-12-13 11.980418919819819 2013-04-03 11.889395669634187 2013-08-02 7.97737779072696 2013-12-26 11.65001285150257 2012-06-29 8.210212142500598 2013-07-16 8.02194925904599 2012-05-18 8.925872590231952 2012-07-14 8.58583751669502 2013-04-04 12.15311162934478 2014-01-09 11.115780183379668 2014-02-15 11.480048853880175 2012-09-15 8.389795211174125 2013-06-16 8.820225056897202 2013-07-17 8.073020528663756 2014-02-25 10.292907089719249 2012-06-09 8.553959536778352 2013-01-29 11.584287970413225 2011-12-25 12.931978548793568 2013-02-03 13.278805295017476 2013-07-23 8.224373097279297 2013-10-14 10.020677712965115 2012-04-22 10.691102166723743 2012-08-26 8.074046177674223 2012-04-14 10.279653898723115 2012-02-04 15.177477698726115 2011-12-12 11.960700933644858 2013-04-08 10.562803028088203 2012-02-24 11.55654339798742 2012-05-24 7.963347222576833 2012-01-25 12.56977395172414 2013-03-22 12.031831346463136 2014-02-17 10.97335237270716 2012-07-23 7.9793680971112995 2013-07-11 7.768358010668186 2012-02-05 15.937699046496816 2012-12-11 11.74944194859438 2013-01-03 11.358292276207907 2013-02-24 13.977933823863212 2013-05-30 9.037915960090448 2013-07-24 7.881485909691547 2013-09-30 8.837252367685165 2012-08-19 8.390505614770076 2012-12-09 13.404829166362472 2012-12-26 11.648737968636775 2013-08-16 7.85339361477729 2012-03-22 10.205075633773086 2012-04-20 10.30783939890237 2013-06-15 8.647237006615004 2012-06-14 8.14960202646218 2012-07-28 8.082364060152932 2012-09-18 8.473907653972544 2012-10-24 9.76565267233216 2014-02-14 11.812373245988773 2012-03-23 10.11830067717206 2013-04-06 11.070999065533346 2013-08-03 7.737711831488549 2012-01-24 12.89820547886497 2013-10-26 9.610571373023888 2012-01-19 12.266832239651416 2012-11-30 12.270830474051094 2012-07-03 8.331989651411101 2012-05-07 10.517486578440565 2011-12-06 11.673416666666668 2012-09-30 10.132886537908105 2013-04-14 10.049026186251407 2013-05-06 8.637916963554385 2013-11-28 11.040900512190326 2014-01-28 11.342741782546321 2012-07-17 8.142612848457777 2012-08-13 8.063635246769753 2013-10-13 10.965290696162793 2014-01-08 11.001265875014832 2014-01-19 10.624456210429939 2012-05-28 8.104686320120205 2012-02-15 13.067887341446452 2012-10-23 10.07007248793103 2012-12-07 12.792815809090918 2012-06-24 9.000452196281987 2012-04-23 10.251355715205328 2012-01-17 13.283576576126126 2012-07-07 8.441053270375367 2013-03-05 10.831471691925008 2013-07-31 7.96007453362562 2014-01-03 11.488910583719095 2012-06-22 8.455070269542041 2012-05-20 9.793731367595594 2012-01-22 13.730465160860655 2012-11-07 11.225314394428208 2014-02-24 10.411359192686868 2012-11-03 11.380744652239079 2013-09-28 8.990865816183295 2013-11-14 10.630769606090674 2013-12-15 12.244833760414616 2011-12-10 12.6418099975 2012-01-10 11.969908011320758

<b>`INFERENCE:` </b> The average energy consumption for any household in a day in UK ranges from 0-16 kWh. The maximum average energy in a day is 15.9 kWh and the least is 7.22 kWh.

<b> `5. How much energy on an average does a typical household in UK consume in a day across the year based on the ACORN group. `</b>

In [0]:
'''
Let us see how this trend is distributed among the Acorn groups
'''
gbda = df.groupBy(['day', 'Acorn_grouped']).mean('energy_sum')
display(gbda)

day Acorn_grouped avg(energy_sum) 2013-06-20 Affluent 9.03160066905787 2013-09-24 Affluent 9.526662889058652 2014-02-27 Affluent 12.113709676430807 2012-07-17 Affluent 8.985623008366542 2012-07-27 Affluent 8.924739433565389 2013-02-14 Affluent 14.118078346824293 2013-11-04 Affluent 11.992550000495045 2013-12-11 Affluent 13.729320524317496 2014-02-05 Affluent 13.408866392446043 2012-06-20 Affluent 8.979145150946643 2012-06-30 Affluent 8.911587804856016 2012-07-02 Affluent 9.73448995485491 2012-03-18 Affluent 13.163803193439719 2012-04-26 Affluent 11.290164144816412 2012-01-28 Affluent 14.79761390617761 2013-08-05 Affluent 8.558435411949679 2012-08-24 Affluent 8.504666997369728 2012-12-11 Affluent 13.956794305831039 2013-04-03 Affluent 13.87646676053748 2013-05-19 Affluent 10.134735966555665 2012-02-15 Affluent 14.482636846052634 2012-07-26 Affluent 8.543983590502233 2012-12-01 Affluent 15.365915556723262 2013-11-19 Affluent 13.057376428592738 2014-02-17 Affluent 12.741348140237596 2012-06-06 Affluent 9.484648565605493 2012-04-24 Affluent 11.52260918051512 2012-08-23 Affluent 8.465376673895781 2013-01-20 Affluent 17.59645525834871 2013-04-01 Affluent 13.755975954832628 2013-05-04 Affluent 9.727298859819737 2013-05-28 Affluent 10.234352382000004 2013-08-04 Affluent 8.407099661828738 2013-10-15 Affluent 11.062565131352152 2012-07-30 Affluent 8.755280955842865 2013-04-15 Affluent 10.595606145106382 2013-09-26 Affluent 9.537552980335139 2012-09-26 Affluent 10.656459184693887 2012-10-08 Affluent 11.684017167620413 2013-05-24 Affluent 11.027773438791046 2013-06-30 Affluent 9.124019617894739 2013-10-02 Affluent 10.181925888537547 2012-06-29 Affluent 9.150812570951297 2011-12-24 Affluent 14.155748338410596 2013-04-10 Affluent 11.900739479290783 2013-06-10 Affluent 9.759562081709646 2014-01-14 Affluent 13.404823736016303 2012-08-11 Affluent 8.646838214143928 2012-12-23 Affluent 13.107168200967742 2013-03-01 Affluent 14.079373945782574 2013-04-04 Affluent 14.16773584886792 2014-02-11 Affluent 13.609606994393 2012-06-12 Affluent 10.547812089228009 2012-09-11 Affluent 9.190405551363407 2012-10-11 Affluent 11.540997630047393 2012-10-14 Affluent 12.160580338846888 2012-06-08 Affluent 9.918201716983448 2013-02-16 Affluent 13.093677794483076 2013-07-04 Affluent 8.91533173355705 2014-02-19 Affluent 12.263030553340244 2012-06-23 Affluent 9.542900683675802 2012-09-15 Affluent 9.201261493129014 2012-08-07 Affluent 9.028685857022333 2012-09-02 Affluent 9.903874503373013 2013-03-07 Affluent 12.625639605443466 2013-03-14 Affluent 13.86932972648166 2013-03-16 Affluent 14.016338502397748 2013-08-20 Affluent 8.2587259 2013-08-24 Affluent 8.808398534975614 2013-12-19 Affluent 13.242558929640863 2012-04-03 Affluent 11.220467553097345 2013-02-22 Affluent 15.023550743249542 2013-10-07 Affluent 10.06776804871414 2013-12-31 Affluent 13.511794119117647 2012-05-26 Affluent 8.838269593175676 2012-12-22 Affluent 13.92885668023041 2013-11-08 Affluent 12.20857659851263 2013-12-03 Affluent 13.414676571687659 2014-02-15 Affluent 13.085524275671471 2012-05-24 Affluent 8.883087852985588 2013-06-02 Affluent 9.514046711916109 2013-09-06 Affluent 9.174966224718554 2012-06-09 Affluent 9.519689866300364 2012-05-04 Affluent 11.506806026647835 2012-08-06 Affluent 9.018756326253104 2012-09-09 Affluent 9.29462252003968 2012-12-10 Affluent 14.856056956270097 2012-07-04 Affluent 9.142892333821868 2012-05-09 Affluent 10.943249338196287 2013-08-17 Affluent 8.60661605815085 2013-10-20 Affluent 11.225104661755953 2012-06-16 Affluent 8.895989165455708 2013-09-17 Affluent 10.661923379076622 2012-04-19 Affluent 12.296593181851401 2012-12-21 Affluent 13.608868722570246 2013-08-13 Affluent 8.380821932411449 2013-11-25 Affluent 13.649140129784021 2012-08-28 Affluent 8.77649280302729 2013-07-07 Affluent 8.820020681144783 2013-12-26 Affluent 13.09392093258996 2013-12-28 Affluent 12.780009118135762 2014-01-20 Affluent 13.410945862768127 2014-01-28 Affluent 13.355957590853349 2012-06-10 Affluen

<b>`INFERENCE:` </b> It is quite evident from the above visualization that Affluent group tops the group in terms of daily average energy consumption.

In [0]:
df = df.withColumn("day", to_date("day", 'yyyy-MM-dd'))

In [0]:
#Importing all blocks from half_hourly_dataset

hh_df_final = pd.DataFrame()

for i in range(0, 50):
    path = 'https://raw.githubusercontent.com/OmkarKatkade/Data-603/main/halfhourly_dataset/halfhourly_dataset/block_' + str(i) +'.csv';
    hh_df = pd.read_csv(path)
    hh_df_final = hh_df_final.append(hh_df)
    
hh_df_final

The python kernel is unresponsive.

In [0]:
'''
Here we are performing Data Cleaning and Data Manipulation for the Half Hourly Dataset.
Replacing all null values in the column 'energy(kWh)' with 0. There aren't many such instances though.
'''
hh_df_final["energy(kWh/hh)"] = hh_df_final["energy(kWh/hh)"].fillna(0)
hh_df_final["energy(kWh/hh)"] = hh_df_final["energy(kWh/hh)"].replace("Null", 0)
hh_df_final["energy(kWh/hh)"] = pd.to_numeric(hh_df_final["energy(kWh/hh)"])

'''
Converting the data type of the column 'date' from string to date.
'''
hh_df_final["date"] = pd.to_datetime(hh_df_final['tstp']).dt.date

'''
Creating a dataframe
'''
hh_df = spark.createDataFrame(hh_df_final)
hh_df.printSchema()

'''
Converting the column 'tstp' from string type to timestamp and
also extracting the hour from the timestamp in a different column
'''
hh_df = hh_df.withColumn("datetype_timestamp", to_timestamp(col("tstp")))
hh_df = hh_df.withColumn("hour", hour(col("tstp")))


In [0]:
hh_df.show(10)

<b> ` 6. Which hours of the day consumes the most and the least energy?`</b>

In [0]:
'''
Since we have information on the half hourly energy consumption,
we are curious to know which hours of the day consumes the most energy.
'''
dfrt = hh_df.groupBy("hour").sum("energy(kWh/hh)").withColumnRenamed('sum(energy(kWh/hh))', 'Energy_Sum')
display(dfrt)

<b>`INFERENCE:` </b> The energy consumption increases drastically from 15th hour (3 p.m) and reaches the peak at 19th hour (7 p.m).

In [0]:
'''
Extracting date from the datetime in the half hourly dataset.
'''
hh_df_final["date"] = pd.to_datetime(hh_df_final['tstp']).dt.date

In [0]:
#Importing weather daily data
weather = pd.read_csv("https://raw.githubusercontent.com/OmkarKatkade/Data-603/main/weather_daily_darksky.csv")
weather["date"] = pd.to_datetime(weather['temperatureMaxTime'])
weather_df = spark.createDataFrame(weather)
weather_df.show()

In [0]:
'''
Extracting only the important features from the weather dataset
'''
weather_df_1 = weather_df.select("date","temperatureMax", "humidity")
wtr = weather_df_1.withColumn("day",to_date("date")).collect()
weather_df_1 = spark.createDataFrame(wtr)
weather_df_1.show()

In [0]:
'''
Joining the important features from the weather dataset with the daily dataset based on the day
'''
df_new = df.join(weather_df_1, ['day'], how='inner')
df_new.show()

In [0]:
'''
Here the aim is to analyze the impact of Temperature and Humidity on the energy consumption.
'''

df_grped_by_day = df_new.select("day", "temperatureMax", "month", "energy_sum").groupBy("day").sum("energy_sum")
df_grped_for_hum = df_new.select("day","humidity","month","energy_sum").groupBy("day").sum("energy_sum")

df_grped_by_day1 = df_new.select("day", "temperatureMax").groupBy("day").mean("temperatureMax")
df_grped_for_hum1 = df_new.select("day", "humidity").groupBy("day").mean("humidity")

df_grped_by_day = df_grped_by_day.join(df_grped_by_day1, ["day"])
df_grped_for_hum = df_grped_for_hum.join(df_grped_for_hum1, ["day"])

df_grped_by_day = df_grped_by_day.withColumn('dayOfWeek', dayofweek(col('day')))
df_grped_by_day = df_grped_by_day.withColumn('month', month(col('day')))
df_grped_by_day = df_grped_by_day.withColumn('year', year(col('day')))

df_grped_for_hum = df_grped_for_hum.withColumn('dayOfWeek', dayofweek(col('day')))
df_grped_for_hum = df_grped_for_hum.withColumn('month', month(col('day')))
df_grped_for_hum = df_grped_for_hum.withColumn('year', year(col('day')))

df_temp_temp = df_grped_by_day.groupBy("month").avg("avg(temperatureMax)").withColumnRenamed("avg(avg(temperatureMax))", "Temperature")
df_temp2_temp = df_grped_by_day.groupBy("month").avg("sum(energy_sum)").withColumnRenamed("avg(sum(energy_sum))", "Avg_Energy_Sum")
df_temp3_temp = df_temp_temp.join(df_temp2_temp, ["month"], how="inner")

df_temp = df_grped_for_hum.groupBy("month").avg("avg(humidity)").withColumnRenamed("avg(avg(humidity))", "Humidity")
df_temp2 = df_grped_for_hum.groupBy("month").avg("sum(energy_sum)").withColumnRenamed("avg(sum(energy_sum))", "Avg_Energy_Sum")
df_temp3 = df_temp.join(df_temp2, ["month"], how="inner")

pandasDF = df_temp3_temp.toPandas()
pandasDF1 = df_temp3.toPandas()

<b> ` 7. What is the impact of Humidity on the energy consumption?`</b>

In [0]:
'''
Visualizing the impact of humidity on the energy consumption
'''
sns.set(rc={'figure.figsize':(11.7,8.27)})
ax = sns.lineplot(x = pandasDF1['month'], y = pandasDF1['Humidity'], legend= True);
ax2 = plt.twinx()
sns.lineplot(x = pandasDF1['month'], y = pandasDF1['Avg_Energy_Sum'], color="g", ax=ax2, legend = True)
plt.legend(labels=["Avg_Energy_Sum"])
plt.show()


<b>`INFERENCE:` </b> From the above visualization, we can see that energy consumption is directly proportional to the humidity for the household in the UK.

<b> ` 8. What is the impact of Temperature on the energy consumption?`</b>

In [0]:
'''
Visualizing the impact of temperature on the energy consumption
'''
sns.set(rc={'figure.figsize':(11.7,8.27)})
ax = sns.lineplot(x = pandasDF['month'], y = pandasDF['Temperature'], legend= 'auto');
ax2 = plt.twinx()
sns.lineplot(x = pandasDF['month'], y = pandasDF['Avg_Energy_Sum'], color="g", ax=ax2, legend = 'auto')
plt.legend(labels=["Avg_Energy_Sum"])
plt.show()

<b>`INFERENCE:` </b> From the visualization above, we can say that as the energy consumption increases with the decrease in temperature.